### PR TITLE
Improve timer precision on Linux by timerfd

### LINKED ISSUE #81: Improve timer precision on Linux
Unfortunately, epoll only supports blocking with timeouts of millisecond precision (because `epoll_wait()` accepts milliseconds as an argument), which means our timers will therefore also have millisecond precision.

The solution to improving precision is: create a timerfd that fires whenever the timeout is, thus interrupting `epoll_wait()`.

### PR DESCRIPTION
**Pull request by wusyong**
As #81 mentioned, this PR creates a timerfd that starts a timer to wait for timeout instead of waiting on `epoll_wait`. After timeout is over, it will call `epoll_wait` to return immediately.

Close #81

### CODE REVIEW

**Comment by wusyong** [comment]
Looks like timerfd is missing on android target ([#1589](https://github.com/rust-lang/libc/issues/1589)). I think I could add temporary ffi module to fix it? Is this good for you?

File: src/reactor.rs [hunk 1]
```diff
@@ -403,9 +407,28 @@ mod sys {
...
+                        Lazy::new(|| unsafe { timerfd_create(CLOCK_REALTIME, TFD_CLOEXEC) });
+                    let timespec = *TimeSpec::nanoseconds(t.as_nanos() as i64).as_ref();
+                    let new_value = itimerspec {
+                        it_interval: timespec,
+                        it_value: timespec,
+                    };
+                    // Start a relative timer
+                    let res = timerfd_settime(*TFD, 0, &new_value, std::ptr::null_mut());
+                    Errno::result(res).map(|r| r as usize).map_err(io_err)?;
+                    // TFD_NONBLOCK isn't set, so this will wait for the first timeout.
```
Comment (polachok) [inline src/reactor.rs:425]: Why would you do it like that instead of registering timerfd with epoll? Seems like it would block any other events.

Comment (wusyong) [reply src/reactor.rs:425]: If it is registered to epoll, doesn't mean the precision is still millisecond? And isn't this method the reason to block I/O events? (https://github.com/stjepang/smol/blob/master/src/reactor.rs#L261)

Comment (polachok) [reply src/reactor.rs:425]: >If it is registered to epoll, doesn't it mean the precision is still millisecond?

No. The kernel will return from `epoll_wait` as soon as the timer fires if it is in the epoll set.

>And isn't this method the reason to block I/O events?

Yes, it blocks waiting on *multiple* i/o sources waiting for events. You block on one source (timer), which means that other events won't be processed until timer fires.

Comment (wusyong) [reply src/reactor.rs:425]: Ah thanks for clarify. I think I'll close it for now to fix this and also the android target issue. Will reopen it once everything is resolved.

**Pull request closed without merging**